ControlNet backend first draft

GreggHelt2 opened the pull request:
This is a working but rough draft for adding ControlNet support to the InvokeAI backend. 
All code changes are down in invokeai.backend.generator and invokeai.backend.stable_diffusion.diffusion. 
Example of usage is in scripts/invokeai_controlnet_example.py
My basic strategy is to incorporate down in invokeai.backend.generator Generator, Txt2Img, etc. so usage in the "AI Core" looks like:
`txt2img_canny = Txt2Img(model, control_model=canny_controlnet)`
`outputs = txt2img_canny.generate(prompt="insert generic prompt", control_image=canny_image)`

I've only implemented for Txt2Img, but most of the affected code is shared by Img2Img, Inpaint, etc. as well. And the small account of code specific to Txt2Img could be copied for other Generators subclasses, or probably be pushed down into the Generator base class with a little refactoring. But I wanted to get some feedback before going any further.

Node support is not included in the current PR, but should be one of the next steps worked on.  My understanding is that building this into new Generator framework should hopefully make Node support straightforward.

damian0815 commented:
so `generate_uc_and_c_and_ec` is in desperate need of refactoring. i see you wrote a comment about it - any thoughts on what form a refactor could take to make things better for this PR?

GreggHelt2 commented:
@damian0815 
Oops that's a comment I accidentally left in. At one point I was having trouble trying to get ControlNetModel to take the results coming out of get_uc_and_c_and_ec(). So I was trying  a seperate call to compel() to feed the ControlNetModel, hence the comment about having to make double compel calls. But after looking more into get_uc_and_c_and_ec() function I figured out how to wrap the results for ControlNet. I'll strip out that comment. And I'll take a closer look at the function and try to give feedback.

damian0815 commented:
ahh nice that this resolved itself! don't worry too much, i have no concrete ideas myself.

cinemapara commented:
@GreggHelt2 does it also support controlnet-pose ?

GreggHelt2 commented:
> @GreggHelt2 does it also support controlnet-pose ?

Yes, I've tested with openpose input. Also, this implementation _should_ work with any of the ControlNet models hosted at https://huggingface.co/lllyasviel/ControlNet, though I haven't tested them all yet.

artsparkAI commented:
Hi there, I run a forked version of InvokeAI at artspark.io that has multi-controlnet support. However its not directly mergeable as it was a bit of a hack as I implemented it before MultiControlnet was merged into diffusers. I'll take a look at this PR and compare to my implementation to see how to get it in.

My first comment though is that it looks like Diffusers now has a [StableDiffusionControlNetPipeline](https://github.com/huggingface/diffusers/pull/2627), any reason we shouldn't use that?

GreggHelt2 commented:
I've added MultiControlNet support. I've been testing image generation using both a Canny ControlNet model and an OpenPose ControlNet model at the same time, and it's working well.
But I seem to be stuck in a rebase loop. I push to the InvokeAI rep, on the feat/controlnet_backend_draft branch, and it says I'm out of date with the main branch. If

GreggHelt2 commented:
> Hi there, I run a forked version of InvokeAI at artspark.io that has multi-controlnet support. However its not directly mergeable as it was a bit of a hack as I implemented it before MultiControlnet was merged into diffusers. I'll take a look at this PR and compare to my implementation to see how to get it in.

I just added MultiControlNet support to the PR. Please take a look and let me know what you think!
> 
> My first comment though is that it looks like Diffusers now has a [StableDiffusionControlNetPipeline](https://github.com/huggingface/diffusers/pull/2627), any reason we shouldn't use that?

So for generative pipelines the general strategy for InvokeAI is currently to use core diffusers models, schedulers, and other lower level classes, but to put those together in custom InvokeAI pipelines. This is in keeping with the [diffusers philosophy](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy):

> (Diffusers) Pipelines are designed to be easy to use..., are not feature complete, and should loosely be seen as examples of how to use [models](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy#models) and [schedulers](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy#schedulers) for inference.

So for the ContrrolNet implementation in this PR, I started with the recently released diffusers StableDiffusionControlNetPipeline and tried to weave the ControlNet functionality from that into the more "feature complete" InvokeAI Generator framework. Part of my expectation is that this strategy for txt2img ControlNet support will make it easier to provide ControlNet capabilities for other tasks like img2img, inpainting, etc. without excessive code duplication.

And please consider the above non-authoritative -- I'm fairly new to the project, and doing my best to integrate my contributions into the codebase. Anyone with a better handle on evolution of InvokeAI want to weight in on pipeline design?

artsparkAI commented:
> I've added MultiControlNet support. I've been testing image generation using both a Canny ControlNet model and an OpenPose ControlNet model at the same time, and it's working well.

Amazing!!

> But I seem to be stuck in a bad rebase loop:
> Ugh. Any suggestions? I'm tempted to just start a fresh branch off of main and port my changes...

Yeah that's annoying :/ this is why I personally go for merge instead of rebase. Just do one merge commit for conflicts and you're done.

> So for generative pipelines the general strategy for InvokeAI is currently to use core diffusers models, schedulers, and other lower level classes, but to put those together in custom InvokeAI pipelines that have more bells and whistles. This is in keeping with the [diffusers philosophy](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy):

Yup that totally makes sense. It's only a small bit of code to copy in. I think for this PR we want to figure out, or at least plan for, how to get Img2Img and Canvas modes to also be able to use the controlnet models, as I think that will be the actual interface people will use.

 One thing I'll say as I was hacking on this, I found myself really wishing that Invoke had a capability to swap in arbitrary pipelines from diffusers, so I could get InstructPix2Pix or [Long Prompt pipeline]([Long Prompt Weighting Stable Diffusion](https://huggingface.co/docs/diffusers/using-diffusers/custom_pipeline_examples#long-prompt-weighting-stable-diffusion) easily running.
 
Same goes for me, I'm not completely familiar with the codebase yet so definitely need to defer to longer term contributors on anything.

artsparkAI commented on invokeai/backend/generator/txt2img.py:
```diff
@@ -61,13 +75,34 @@ def get_make_image(
             ),
         ).add_scheduler_args_if_applicable(pipeline.scheduler, eta=ddim_eta)
 
+        if control_image is not None:
```
More terse, unless there was a reason not to
```
            if isinstance(self.control_model, ControlNetModel):
                control_image = pipeline.prepare_control_image(image=control_image, do_classifier_free_guidance=do_classifier_free_guidance)
            elif isinstance(self.control_model, MultiControlNetModel):
                control_image = [pipeline.prepare_control_image(image=image_, do_classifier_free_guidance=do_classifier_free_guidance) for image_ in control_image]
```

artsparkAI commented:
Looks awesome! I need to take a look at my implementation to remember how I handled Img2Img and Canvas, but other than that really great that it's already working.

hipsterusername commented:
> One thing I'll say as I was hacking on this, I found myself really wishing that Invoke had a capability to swap in arbitrary diffusers pipelines, or components from them, so I could get InstructPix2Pix or [Long Prompt Weighting Stable Diffusion](https://huggingface.co/docs/diffusers/using-diffusers/custom_pipeline_examples#long-prompt-weighting-stable-diffusion) easily running.

This is precisely what we're aiming to get to with our backend rewrite (Nodes)

GreggHelt2 commented:
This PR is now superseded by PR #3156 
Moved to a new branch feat/controlnet_backend to escape rebase hell on previous brach feat/controlnet_backend_draft.
